keps/sig-api-machinery/2885-server-side-unknown-field-validation/README.md

```diff
@@ -0,0 +1,916 @@
+-->
+### Dependencies
+
+This section must be completed when targeting beta to a release.
+
+N/A
+
+<!--
+###### Does this feature depend on any specific services running in the cluster?
+
+Think about both cluster-level services (e.g. metrics-server) as well
+as node-level agents (e.g. specific version of CRI). Focus on external or
+optional services that are needed. For example, if this feature depends on
+a cloud provider API, or upon an external software-defined storage or network
+control plane.
+
+For each of these, fill in the following—thinking about running existing user workloads
+and creating new ones, as well as about cluster-level services (e.g. DNS):
+  - [Dependency name]
+    - Usage description:
+      - Impact of its outage on the feature:
+      - Impact of its degraded performance or high-error rates on the feature:
+-->
+
+### Scalability
+
+<!--
+For alpha, this section is encouraged: reviewers should consider these questions
+and attempt to answer them.
+
+For beta, this section is required: reviewers must answer these questions.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+###### Will enabling / using this feature result in any new API calls?
+
+Describe them, providing:
+  - API call type (e.g. PATCH pods)
+  - estimated throughput
+  - originating component(s) (e.g. Kubelet, Feature-X-controller)
+Focusing mostly on:
+  - components listing and/or watching resources they didn't before
+  - API calls that may be triggered by changes of some Kubernetes resources
+    (e.g. update of object X triggers new updates of object Y)
+  - periodic API calls to reconcile state (e.g. periodic fetching state,
+    heartbeats, leader election, etc.)
+
+###### Will enabling / using this feature result in introducing new API types?
+
+Describe them, providing:
+  - API type
+  - Supported number of objects per cluster
+  - Supported number of objects per namespace (for namespace-scoped objects)
+
+###### Will enabling / using this feature result in any new calls to the cloud provider?
+
+Describe them, providing:
+  - Which API(s):
+  - Estimated increase:
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+Describe them, providing:
+  - API type(s):
+  - Estimated increase in size: (e.g., new annotation of size 32B)
+  - Estimated amount of new objects: (e.g., new Object X for every existing Pod)
+-->
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+<!--
+Look at the [existing SLIs/SLOs].
+
+Think about adding additional work or introducing new steps in between
+(e.g. need to do X to start a container), etc. Please describe the details.
+
+[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
+-->
+Mutating API calls that opt-in to validation will be slower ([initial
```

Comment:
@wojtek-t I don't think these numbers block an alpha. But I suspect many (most?) clients will start requesting this kind of validation.

Reply:
My understanding is that existing users of client-side validation are all "human-readable" (i.e. kubectl) and so the performance increase for feature parity is negligible. 

If clients that do not currently use schema validation (such as controllers) want to opt-in they do so knowing the performance tradeoffs and can simply continue to not have schema validation. Is this correct?

Reply:
i want to look deeper into proposal, but I won't get to it by Monday.

But from the first glance, I think that 20-30% increase is acceptable, because I think that we actually don't do that for many/most clients.
My personal expectation would be to use that for client-originating clients (kubectl in particular), but given that majority of load is coming from our controller/kubelets/... I don't think we really need to enable it there.

But let me take a deeper look on it early next week - but certainly doesn't look like as alpha blocker.

Reply:
@deads2k - do you agree with the above or you have different expectations?

Reply:
> @deads2k - do you agree with the above or you have different expectations?

The explanation makes sense to me.  To get the feature, I'm willing to pay a performance penalty.  Just wanted to let you know it was coming.